---
title: "DRP-HCB Training Courses"
format:
  html:
    toc: true
    toc-depth: 2
---

## Training courses

### Data science for biology (2026)

![](images/DSB_banner.png)

Learn data science skills with biological datasets. This course introduces the R programming language. R is an extremely powerful tool for data science, statistical analysis and data visualisation. It is favoured by biologists due to its extensive library of packages for bioinformatics and genomic analysis.

The Data Science for Biology course is divided into six modules. Participants can register for the full course or individual modules.

-   Introduction to R and RStudio - 17th February
-   Data manipulation and visualisation (Tidyverse) - 17th March
-   Genomic datasets in R (Bioconductor) - 21st April
-   Introduction to statistics (RStatix) - 12th May
-   Differential expression analysis (DESeq2) - 9th June
-   Data publication and presentation (Quarto and Shiny) - 23rd June

Each module consists of:

-   A full day in person workshop
-   Online training material for self study
-   A practical exercise that can be tailored to your own data
-   Access to monthly bioinformatics drop in sessions

#### Cost

-   All modules: **£200**
-   Individual modules: **£50** 

#### Registration

Please complete this registration form to sign up for the course or individual modules:\
[Data Science for Biology Registration Form](https://forms.office.com/e/hx73VvSAJs)

This course is hosted by The Bioinformatics Core at the Discovery Research Platform for Hidden Cell Biology (DRP-HCB).

Contact [shaun.webb\@ed.ac.uk](mailto:shaun.webb@ed.ac.uk) for more information.

#### Modules

**Introduction to R and RStudio**\
17th February, 10am - 4pm\
Rowan teaching room, The Nucleus, Kings Buildings

Learn the basics of the [R programming language](https://www.r-project.org/) and the [RStudio](https://posit.co/products/open-source/rstudio/) programming environment.

-   Install R, RStudio and packages
-   Use the RStudio interface
-   Basic R syntax
-   Data types and structures
-   Importing and exporting data
-   Plots and statistics with base R

**Data manipulation and visualisation (Tidyverse)**\
17th March, 10am - 4pm\
JCMB 3211, Kings Buildings

The [*Tidyverse*](https://tidyverse.org/) is a collection of R packages designed for data science and the preferred utilities for many R users. This module introduces the core Tidyverse packages for data manipulation and visualisation.

-   Introduction to the Tidyverse packages (tidyr, readr, dplyr, ggplot2)
-   Work with large datasets
-   Import, format, filter and summarise tables of data
-   Create publication ready visualisations

**Genomic datasets in R (Bioconductor packages)**\
21st April, 10am - 4pm\
Rowan teaching room, The Nucleus, Kings Buildings

[Bioconductor](https://www.bioconductor.org/) is a collection of R packages specifically designed for analysing genomic datasets. This module introduces some of the core Bioconductor packages and data formats for working with biological data.

-   Bioconductor overview
-   Importing and exporting genomic data
-   Operations on genomic datasets
-   Plotting genomic data

**Introduction to Statistics (RStatix)**\
12th May, 10am - 4pm\
JCMB 4325, Kings Buildings

An introduction to statistical analysis in R using the [RStatix](https://rpkgs.datanovia.com/rstatix/) package. This module covers common statistical tests and how to apply them to biological datasets.

-   Data exploration and visualisation
-   Common statistical tests (t-test, ANOVA, chi-squared test)
-   Assumptions and diagnostics
-   Post-hoc testing and multiple testing correction

**Differential expression analysis (DESeq2)**\
9th June, 10am - 4pm\
Peter Wilson G155, Kings Buildings

Perform a full differential expression analysis of RNA-seq data using the [DESeq2](https://bioconductor.posit.co/packages/3.19/bioc/html/DESeq2.html) Bioconductor package.

-   Data import and quality control
-   Normalisation and transformation
-   Differential expression testing
-   Visualisation and interpretation of results
-   Downstream analysis (pathway analysis, gene ontology)

**Data publication and presentation (Quarto and Shiny)**\
23rd June, 10am - 4pm\
Peter Wilson G155, Kings Buildings

Learn how to build reproducible analyses and present your results as web pages and interactive applications.

-   Project version control with [git](https://git-scm.com/) and [GitHub](https://github.com/)
-   Package management with [renv](https://rstudio.github.io/renv/)
-   Create reproducible documents and websites with [Quarto](https://quarto.org/)
-   Build interactive data visualisations with [Shiny](https://shiny.posit.co/)

### Analysing high-throughput sequencing data (2026)

Coming soon.

## Data Drop-Ins

Monthly drop-in sessions to get help with bioinformatics analysis and training exercises.

Upcoming dates:

-   Thursday 5th March 2026, 14:00 - 15:00, Swann 7.14, Kings Buildings

-   Thursday 2nd April 2026, 14:00 - 15:00, Swann 7.14, Kings Buildings

-   Friday 1st May 2026, 14:00 - 15:00, Swann 7.14, Kings Buildings

-   Thursday 28th May 2026, 14:00 - 15:00, Swann 7.14, Kings Buildings

-   Friday 25th June 2026, 14:00 - 15:00, Swann 7.14, Kings Buildings

## Bioinformatics Live

Regular live demonstration sessions covering a variety of bioinformatics tools.

See the [Bioinformatics Live](bioinformatics_live.html) page for more details and upcoming sessions.
