<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.554">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>DRP-HCB-Bioinformatics - Bioinformatics Live</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">DRP-HCB-Bioinformatics</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./training.html"> 
<span class="menu-text">Training Events</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./training_resources.html"> 
<span class="menu-text">Training Resources</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./best_practices.html"> 
<span class="menu-text">Best Practices</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./bioinformatics_live.html" aria-current="page"> 
<span class="menu-text">Bioinformatics Live</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#upcoming" id="toc-upcoming" class="nav-link active" data-scroll-target="#upcoming">Upcoming</a></li>
  <li><a href="#completed" id="toc-completed" class="nav-link" data-scroll-target="#completed">Completed</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Bioinformatics Live</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>Bioinformatics Live is our series of demonstrations focusing on specific applications of bioinformatics, data analysis and programming. The aim of these sessions is to share knowledge, develop skills and create a network of computational biologists.</p>
<p>You can find the schedule below as well as links to previous talks. We are always keen to have new speakers so please get in touch if you are interested!</p>
<section id="upcoming" class="level3">
<h3 class="anchored" data-anchor-id="upcoming">Upcoming</h3>
<p>Coming soon in 2026.</p>
</section>
<section id="completed" class="level3">
<h3 class="anchored" data-anchor-id="completed">Completed</h3>
<section id="shiny-interactive-apps-for-exploratory-data-analysis" class="level4">
<h4 class="anchored" data-anchor-id="shiny-interactive-apps-for-exploratory-data-analysis"><strong>Shiny: Interactive apps for exploratory data analysis</strong></h4>
<ul>
<li><p><a href="https://docs.google.com/presentation/d/1qYye9iWty6flrjcqqgc-wwH08gIbrwalcubx_z0fmFE/edit?usp=sharing">Presentation slides</a></p></li>
<li><p><a href="https://shiny.posit.co/r/getstarted/shiny-basics/lesson1/index.html">Getting started with Shiny</a></p></li>
<li><p><a href="https://shiny.posit.co/r/gallery/">Shiny gallery</a></p></li>
<li><p>GeneTonic RNA-seq Shiny App</p>
<ul>
<li><p><a href="https://federicomarini.github.io/GeneTonic/articles/GeneTonic_manual.html" class="uri">https://federicomarini.github.io/GeneTonic/articles/GeneTonic_manual.html</a></p></li>
<li><p><a href="http://shiny.imbei.uni-mainz.de:3838/GeneTonic/" class="uri">http://shiny.imbei.uni-mainz.de:3838/GeneTonic/</a></p></li>
</ul></li>
<li><p>Shiny app deployment:</p>
<ul>
<li><p><a href="https://www.shinyapps.io/">Shinyapps IO</a></p></li>
<li><p>Hugging Face</p>
<ul>
<li><p><a href="https://huggingface.co/" class="uri">https://huggingface.co/</a></p></li>
<li><p><a href="https://huggingface.co/docs/hub/spaces-sdks-docker-shiny#shiny-for-r" class="uri">https://huggingface.co/docs/hub/spaces-sdks-docker-shiny#shiny-for-r</a></p></li>
</ul></li>
</ul></li>
</ul>
</section>
<section id="hic-visualisation" class="level4">
<h4 class="anchored" data-anchor-id="hic-visualisation"><strong>HiC visualisation</strong></h4>
<ul>
<li><p><a href="https://uoe-my.sharepoint.com/:p:/g/personal/swebb1_ed_ac_uk/Ead91ODqVJNHuQ454_435f8BTjVZM30Ss9qRic5WhioAlQ?e=2JRPcr">Presentation slides</a></p></li>
<li><p>A brief overview of available tools for visualising HiC data.</p></li>
<li><p>The pros and cons of different tools both for genome wide exploration and plotting at specific viewpoints.</p></li>
<li><p><a href="https://higlass.io/">HiGlass</a>&nbsp;vs <a href="https://www.aidenlab.org/juicebox/">Juicebox</a>&nbsp;for Genome wide heatmaps</p>
<ul>
<li>What are the differences?</li>
</ul></li>
<li><p>Plotting with <a href="https://hicexplorer.readthedocs.io/en/latest/">HiCExplorer</a>&nbsp;and <a href="https://cooltools.readthedocs.io/en/latest/">Cooltools</a></p></li>
</ul>
</section>
<section id="simple-unix-commands-to-enhance-your-command-line-experience" class="level4">
<h4 class="anchored" data-anchor-id="simple-unix-commands-to-enhance-your-command-line-experience"><strong>Simple Unix commands to enhance your command line experience:</strong></h4>
<ul>
<li><p><a href="https://docs.google.com/presentation/d/157U6XQoqcehqGXHKs6D847wwPaM3CUujml2f5dPMNmI/edit?usp=sharing">Presentation slides</a></p></li>
<li><p>Input dataset and shell script:&nbsp;<a href="https://bifx-core3.bio.ed.ac.uk/Shaun/Shaun/training/unix_tutorial/" class="uri">https://bifx-core3.bio.ed.ac.uk/Shaun/Shaun/training/unix_tutorial/</a></p></li>
<li><p>Ensembl Biomart:&nbsp;<a href="https://www.ensembl.org/info/data/biomart/index.html" class="uri">https://www.ensembl.org/info/data/biomart/index.html</a></p></li>
<li><p>Notes on regular expressions:&nbsp;<a href="https://www.datacamp.com/cheat-sheet/regular-expresso" class="uri">https://www.datacamp.com/cheat-sheet/regular-expresso</a></p></li>
</ul>
</section>
<section id="analysing-hts-data-with-deeptools" class="level4">
<h4 class="anchored" data-anchor-id="analysing-hts-data-with-deeptools"><strong>Analysing HTS data with deepTools</strong></h4>
<ul>
<li><p><a href="https://docs.google.com/presentation/d/11TwobnfGEQkqSQ6_drh0gf5FRWTYcuCk78KHOXTxLGw/edit?usp=sharing">Presentation slides</a></p></li>
<li><p><a href="https://deeptools.readthedocs.io/en/stable/index.html">DeepTools documentation</a></p></li>
<li><p><a href="https://www.bioconductor.org/packages/release/bioc/vignettes/profileplyr/inst/doc/profileplyr.html">ProfilePlyr</a>: Importing deepTools objects into R</p></li>
<li><p>Example <a href="https://bifx-core3.bio.ed.ac.uk/Shaun/Shaun/deeptools/deepTools.sh">deepTools shell script</a>&nbsp;</p></li>
<li><p>Example <a href="https://bifx-core3.bio.ed.ac.uk/Shaun/Shaun/deeptools/profilePlyr.R">profilePlyr R script</a></p></li>
</ul>
</section>
<section id="sharing-and-publishing-completely-reproducible-data-analyses" class="level4">
<h4 class="anchored" data-anchor-id="sharing-and-publishing-completely-reproducible-data-analyses"><strong>Sharing and publishing completely reproducible data analyses​​​​​​​​​​​​​​</strong></h4>
<ul>
<li><p><a href="https://uoe-my.sharepoint.com/:p:/g/personal/swebb1_ed_ac_uk/ETB-ZQHEQ41HpByvRVKEEVkBg5kdVHfCzCc-8XJN90mtKQ?e=bcrZRW">Presentation slides</a></p></li>
<li><p>Publishing raw data with&nbsp;<a href="https://www.ncbi.nlm.nih.gov/geo/">GEO</a></p></li>
<li><p>Publishing scripts and workflows with&nbsp;<a href="https://github.com/">GitHub</a></p></li>
<li><p>Using&nbsp;<a href="https://docs.sylabs.io/guides/3.5/user-guide/index.html">Singularity</a>&nbsp;containers to reproduce computing environments and software dependencies</p></li>
<li><p>Publishing software containers with&nbsp;<a href="https://zenodo.org/">Zenodo</a>&nbsp;and linking to a GitHub release</p></li>
<li><p>Downloading data and code to reproduce the analysis<strong>​​​​​​​</strong></p></li>
<li><p>Resources</p>
<ul>
<li><p>Snakefile and Singularity definition file:&nbsp;<a href="https://github.com/hyweldd/GSE230112_hyb" class="uri">https://github.com/hyweldd/GSE230112_hyb</a></p></li>
<li><p>GitHub repository snapshot on Zenodo:&nbsp;<a href="https://zenodo.org/record/7875936" class="uri">https://zenodo.org/record/7875936</a></p></li>
<li><p>Pre-print describing the experiment:&nbsp;<a href="https://www.biorxiv.org/content/10.1101/2022.05.20.492787v1" class="uri">https://www.biorxiv.org/content/10.1101/2022.05.20.492787v1</a></p></li>
<li><p>Singularity container in Zenodo (currently private):&nbsp;<a href="https://zenodo.org/record/7875547" class="uri">https://zenodo.org/record/7875547</a></p></li>
<li><p>GEO record for raw data (currently private):&nbsp;<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE230112" class="uri">https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE230112</a></p></li>
</ul></li>
</ul>
</section>
<section id="building-reproducible-workflows-with-snakemake" class="level4">
<h4 class="anchored" data-anchor-id="building-reproducible-workflows-with-snakemake"><strong>Building reproducible workflows with Snakemake</strong></h4>
<ul>
<li><p><a href="https://docs.google.com/presentation/d/1atUB26GZ5j_IldHm-b1az33PGNvw1j0azrFD3_gKOug/edit?usp=sharing">Presentation slides</a></p></li>
<li><p>Introduction to workflow languages</p></li>
<li><p>Why should I write workflows instead of bash scripts</p></li>
<li><p>Snakemake workflows and rules</p></li>
<li><p>Rule parameters and wildcards</p></li>
<li><p>Running a workflow: Outputs and&nbsp;dry runs</p></li>
<li><p><strong>Snakemake tutorial:</strong> <a href="https://snakemake.readthedocs.io/en/stable/tutorial/tutorial.html" class="uri">https://snakemake.readthedocs.io/en/stable/tutorial/tutorial.html</a></p></li>
</ul>
</section>
<section id="fetching-published-sequencing-data-from-geosra" class="level4">
<h4 class="anchored" data-anchor-id="fetching-published-sequencing-data-from-geosra"><strong>Fetching published sequencing data from GEO/SRA</strong></h4>
<ul>
<li><p>Searching and exploring SRA (the Sequence Read Archive)&nbsp;<a href="https://www.ncbi.nlm.nih.gov/sra" class="uri">https://www.ncbi.nlm.nih.gov/sra</a></p></li>
<li><p>Searching published datasets in GEO (Gene Expression Omnibus)&nbsp;<a href="https://www.ncbi.nlm.nih.gov/geo/" class="uri">https://www.ncbi.nlm.nih.gov/geo/</a></p></li>
<li><p>Downloading sequence files with sra-tools&nbsp;<a href="https://github.com/ncbi/sra-tools" class="uri">https://github.com/ncbi/sra-tools</a></p></li>
<li><p>SRA explorer tool&nbsp;<a href="https://sra-explorer.info/" class="uri">https://sra-explorer.info/</a></p></li>
<li><p>ENA (European Nucleotide Archive)&nbsp;<a href="https://www.ebi.ac.uk/ena/" class="uri">https://www.ebi.ac.uk/ena/</a></p></li>
<li><p><strong>VIDEO:</strong> <a href="https://www.youtube.com/watch?v=fAaVzhD7DOU">Fetching sequencing data from SRA</a></p></li>
</ul>
</section>
<section id="a-tour-of-the-ensembl-website-and-databases" class="level4">
<h4 class="anchored" data-anchor-id="a-tour-of-the-ensembl-website-and-databases"><strong>A tour of the Ensembl website and databases</strong></h4>
<p>The <a href="https://www.ensembl.org/index.html">Ensembl</a> genome database is an online resource for genomic&nbsp;annotation, tools and visualisation:</p>
<ul>
<li><p>Genome assemblies for 300+ species</p></li>
<li><p>Gene and transcript annotations</p>
<ul>
<li>Gene/transcript centric data e.g.&nbsp;expression, proteins, homologues, variants</li>
</ul></li>
<li><p>Variation, regulation and comparative genomics data</p></li>
<li><p>Analysis tools such as&nbsp;BLAST and the Variant Effect Predictor</p></li>
<li><p>Data downloads via FTP, BioMart portal and programming APIs for R/Python/Perl</p></li>
<li><p>Interactive genome browser</p></li>
<li><p><strong>VIDEO:</strong> <a href="https://www.youtube.com/watch?v=JoJGg8xd0e0">Quick guide to Ensembl</a></p></li>
<li><p><strong>VIDEO:</strong> <a href="https://www.youtube.com/playlist?list=PLqB8Yx1tGBMb8jn57vpEGIPh1RfTRqQko">Ensembl virtual workshop playlist</a></p></li>
<li><p><strong>TUTORIAL:</strong> <a href="https://www.ebi.ac.uk/training/online/courses/ensembl-browsing-genomes/">Ensembl training course</a></p></li>
</ul>
</section>
<section id="analysing-cutrun-sequencing-data-kashyap-chhatbar" class="level4">
<h4 class="anchored" data-anchor-id="analysing-cutrun-sequencing-data-kashyap-chhatbar"><strong>Analysing CUT&amp;RUN sequencing data (Kashyap Chhatbar)</strong></h4>
<p>CUT&amp;RUN (<a href="https://doi.org/10.7554/eLife.21856">Skene et al.&nbsp;2017</a>) is a method to analyse protein interactions with DNA.&nbsp;Unlike ChIP, which fragments and solubilizes total chromatin, CUT&amp;RUN is performed in situ, allowing for both quantitative high-resolution chromatin mapping and probing of the local chromatin environment. Analysis steps are like ChIP-seq with a few additional steps. The talk will cover accelerated analysis using <a href="https://nf-co.re/cutandrun">NextFlow CUT&amp;RUN pipeline</a> to automate some of the steps below:</p>
<ul>
<li><p>Mapping sequencing reads to host and spike-in genomes</p></li>
<li><p>Calculate normalization factor and generate spike-in normalized genome browser tracks for visualisation</p></li>
<li><p>Analyse fragments by size</p></li>
<li><p>Peak calling using sparse enrichment analysis algorithm SEACR (<a href="https://doi.org/10.1186/s13072-019-0287-4">Meers et al.&nbsp;2019</a>)</p></li>
<li><p>Generate summary plots and heatmaps</p></li>
</ul>
</section>
<section id="genome-browsers-igv-demonstration" class="level4">
<h4 class="anchored" data-anchor-id="genome-browsers-igv-demonstration"><strong>Genome browsers (IGV demonstration)</strong></h4>
<ul>
<li><p><a href="https://docs.google.com/presentation/d/1Avd-X6TnkITGeayfPPc4u7u3O6O4ojpNA3RMenFzl1M/edit?usp=sharing">Presentation slides</a></p></li>
<li><p>Overview of commonly used genome browsers</p>
<ul>
<li>Websites, servers, programming tools, desktop applications</li>
</ul></li>
<li><p>IGV demo</p>
<ul>
<li><p>Loading data and genomes</p></li>
<li><p>Navigating the genome</p></li>
<li><p>Customising the display</p></li>
<li><p>Working with HTS files - bigWig, BAM, VCF, RNA-seq data</p></li>
<li><p><a href="https://www.youtube.com/c/IGVtutorials">IGV YouTube Channel</a></p></li>
</ul></li>
</ul>
</section>
<section id="gene-functional-enrichment-analysis-jose-de-las-heras" class="level4">
<h4 class="anchored" data-anchor-id="gene-functional-enrichment-analysis-jose-de-las-heras"><strong>Gene functional enrichment&nbsp;analysis (Jose de las Heras)</strong></h4>
<ul>
<li><p>Functional enrichment with <a href="https://biit.cs.ut.ee/gprofiler/gost">gProfiler</a> (GO terms, KEGG pathways etc.)</p></li>
<li><p>Visualising and summarising results in R</p></li>
<li><p>Reducing results&nbsp;with <a href="http://revigo.irb.hr/">ReviGo</a> and similarity matrices</p></li>
<li><p>Gene set enrichment analysis (GSEA) with the <a href="https://www.gsea-msigdb.org/gsea/index.jsp">GSEA</a>​​​​​​​ app</p></li>
<li><p><a href="https://uoe-my.sharepoint.com/:p:/g/personal/swebb1_ed_ac_uk/EfvVbyi0aiRPjSSTN1hZtWIBfDOLvZPHszDTBtya2ohykA?e=QyLFxu">Presentation slides</a></p></li>
<li><p><a href="https://www.youtube.com/watch?v=vw85oCOshgg" class="uri">https://www.youtube.com/watch?v=vw85oCOshgg</a></p></li>
</ul>
</section>
<section id="data-analysis-with-r-and-rstudio" class="level4">
<h4 class="anchored" data-anchor-id="data-analysis-with-r-and-rstudio"><strong>Data analysis with R and RStudio</strong></h4>
<ul>
<li><p>Creating projects in RStudio</p></li>
<li><p>Using <a href="https://happygitwithr.com/index.html">git in RStudio</a>​​​​​​​ to version control scripts and software</p></li>
<li><p>Using <a href="https://rstudio.github.io/renv/articles/renv.html">Renv</a> to manage package versions in a project</p></li>
<li><p>Using <a href="https://rmarkdown.rstudio.com/">RMarkdown</a> to annotate and publish your analysis</p></li>
<li><p><a href="https://www.youtube.com/watch?v=TDXCDEFXN1k" class="uri">https://www.youtube.com/watch?v=TDXCDEFXN1k</a></p></li>
<li><p>The <a href="https://usethis.r-lib.org/articles/git-credentials.html">usethis()</a> package in R is particularly useful for managing github within RStudio</p></li>
</ul>
</section>
<section id="managing-your-own-software-with-conda-environments" class="level4">
<h4 class="anchored" data-anchor-id="managing-your-own-software-with-conda-environments"><strong>Managing your own software with Conda environments</strong></h4>
<ul>
<li><p>How to install Conda on a server or your own computer</p></li>
<li><p>The difference between Anaconda and Miniconda</p></li>
<li><p>How to find and install software available via Conda</p></li>
<li><p>How to manage software environments for different projects in Conda</p></li>
<li><p>How to run different versions of the same package</p></li>
<li><p>Sharing environments with other researchers</p></li>
<li><p><a href="https://www.youtube.com/watch?v=F7ak8XANl30" class="uri">https://www.youtube.com/watch?v=F7ak8XANl30</a></p></li>
<li><p><a href="https://wcbinf.github.io/wcb_bioinformatics/software.html#managing-your-own-software-tools-and-packages-with-conda">Install and initialise Conda</a></p></li>
<li><p><a href="https://docs.conda.io/projects/conda/en/latest/user-guide/cheatsheet.html">Conda cheatsheet</a></p></li>
</ul>
</section>
<section id="running-software-on-the-wcb-bioinformatics-servers" class="level4">
<h4 class="anchored" data-anchor-id="running-software-on-the-wcb-bioinformatics-servers"><strong>Running software on the WCB bioinformatics servers</strong></h4>
<ul>
<li><p>Introduction to different types of software on the bioinformatics servers</p></li>
<li><p>Running bioinformatics software via the command line</p></li>
<li><p>Checking which bioinformatics tools are available</p></li>
<li><p>Running specific versions of bioinformatics tools for reproducible research</p></li>
<li><p>Accessing web-based tools hosted on the bioinformatics servers</p></li>
<li><p>Options for managing your own software environment</p></li>
<li><p><a href="https://www.youtube.com/watch?v=h_LzXKm0d_0" class="uri">https://www.youtube.com/watch?v=h_LzXKm0d_0</a></p></li>
<li><p><a href="https://wcbinf.github.io/wcb_bioinformatics/software.html">Further resources</a></p></li>
</ul>
</section>
<section id="a-tour-of-the-wcb-bioinformatics-servers" class="level4">
<h4 class="anchored" data-anchor-id="a-tour-of-the-wcb-bioinformatics-servers"><strong>A tour of the WCB bioinformatics servers</strong></h4>
<ul>
<li><p>Introduction to the bioinformatics core facility and where to find help</p></li>
<li><p>Connecting to the bioinformatics servers (SSH and SSH keys)</p></li>
<li><p>Navigating the file system (home spaces, group spaces, genome reference data etc.)</p></li>
<li><p>Making files available on the web (/homes/www)</p></li>
<li><p>Monitoring running jobs (htop and ps commands)</p></li>
<li><p>Logging out without losing your session (tmux terminal multiplier)</p></li>
<li><p><a href="https://www.youtube.com/watch?v=ZloIeQt_6T4" class="uri">https://www.youtube.com/watch?v=ZloIeQt_6T4</a></p></li>
<li><p><a href="https://wcbinf.github.io/wcb_bioinformatics/customising.html">Further resources</a></p></li>
</ul>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
          // target, if specified
          link.setAttribute("target", "_blank");
          if (link.getAttribute("rel") === null) {
            link.setAttribute("rel", "noopener");
          }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>