<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.554">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>DRP-HCB-Bioinformatics - Additional Training Resources</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">DRP-HCB-Bioinformatics</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./training.html"> 
<span class="menu-text">Training Events</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./training_resources.html" aria-current="page"> 
<span class="menu-text">Training Resources</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./best_practices.html"> 
<span class="menu-text">Best Practices</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./bioinformatics_live.html"> 
<span class="menu-text">Bioinformatics Live</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul class="collapse">
  <li><a href="#linux-command-line" id="toc-linux-command-line" class="nav-link active" data-scroll-target="#linux-command-line">Linux command line</a></li>
  <li><a href="#programming-in-r" id="toc-programming-in-r" class="nav-link" data-scroll-target="#programming-in-r">Programming in R</a></li>
  <li><a href="#programming-in-python" id="toc-programming-in-python" class="nav-link" data-scroll-target="#programming-in-python">Programming in Python</a></li>
  <li><a href="#statistics-for-biology" id="toc-statistics-for-biology" class="nav-link" data-scroll-target="#statistics-for-biology">Statistics for biology</a></li>
  <li><a href="#analysing-hts-experiments" id="toc-analysing-hts-experiments" class="nav-link" data-scroll-target="#analysing-hts-experiments">Analysing HTS experiments</a></li>
  <li><a href="#further-bioinformatics-training-resources" id="toc-further-bioinformatics-training-resources" class="nav-link" data-scroll-target="#further-bioinformatics-training-resources">Further Bioinformatics Training Resources</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Additional Training Resources</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>The DRP-HCB Bioinformatics Core runs annual in-person training workshops covering a variety of topics including Linux command line, analysis of high throughput sequencing (HTS) data, programming languages and data science. Please refer to our <a href="training.html">training</a> page for details of upcoming courses.</p>
<p>This page is intended to provide a set of additional resources for those interested in learning about bioinformatics and acquiring new skills. It is by no means a comprehensive list of online training materials and does not cover everything you need to know to become a skilled bioinformatician. However, it should be enough to get you started and we will happily add more subjects if they are of interest to our users.</p>
<p>Our dedicated bioinformatics servers provide an excellent training platform with an expansive list of tools via the Linux command line, an RStudio server and access to programming languages and libraries such as Python and Perl. If you require an account, please contact <a href="mailto:shaun.webb@ed.ac.uk">Shaun Webb</a>.</p>
<section id="linux-command-line" class="level2">
<h2 class="anchored" data-anchor-id="linux-command-line">Linux command line</h2>
<p>Linux operating systems are the workhorse for most large scale genomics applications and are well suited to managing and processing huge data files produced by high throughput sequencing technologies. Although command-line computing can be daunting to new users, grasping a few fundamental commands and learning to run software packages and pipelines from the terminal will open new doors</p>
<section id="beginner" class="level4">
<h4 class="anchored" data-anchor-id="beginner">Beginner</h4>
<ul>
<li><p><a href="https://files.fosswire.com/2007/08/fwunixref.pdf">Command line cheat-sheet</a></p></li>
<li><p><a href="http://swcarpentry.github.io/shell-novice/">The Unix shell</a></p></li>
<li><p><a href="https://datacarpentry.org/shell-genomics/">Command line for Genomics</a></p></li>
</ul>
</section>
<section id="intermediate" class="level4">
<h4 class="anchored" data-anchor-id="intermediate">Intermediate</h4>
<ul>
<li><p><a href="https://nbis-reproducible-research.readthedocs.io/">Reproducible research</a></p>
<ul>
<li>git, Conda, Snakemake, R Markdown, Jupyter, Docker and Singularity</li>
</ul></li>
<li><p>Package management with Conda</p>
<ul>
<li><p><a href="https://docs.conda.io/projects/conda/en/latest/user-guide/getting-started.html">Getting started with Conda</a></p></li>
<li><p><a href="https://docs.conda.io/projects/conda/en/latest/_downloads/843d9e0198f2a193a3484886fa28163c/conda-cheatsheet.pdf">Conda cheatsheet</a></p></li>
</ul></li>
</ul>
</section>
</section>
<section id="programming-in-r" class="level2">
<h2 class="anchored" data-anchor-id="programming-in-r">Programming in R</h2>
<p>R is a programming language designed for data retrieval, manipulation and visualisation as well as statistical analysis. Learning R will allow you to move beyond spreadsheets and access the full analysis potential of large datasets. You can build multi-dimensional visualisations as well as interactive web applications and documents. It’s easy to get to grips with the basics of R and the <a href="https://rstudio.com/">RStudio</a> interface provides an intuitive graphical environment in which to develop your code. The extensive <a href="https://www.bioconductor.org/">Bioconductor</a> libraries also provide pre-built functions to analyse and interpret all sorts of genomic datasets.</p>
<section id="beginner-1" class="level4">
<h4 class="anchored" data-anchor-id="beginner-1">Beginner</h4>
<ul>
<li><a href="https://rstudio.com/resources/cheatsheets/">R Cheatsheets</a></li>
</ul>
</section>
<section id="beginner---intermediate" class="level4">
<h4 class="anchored" data-anchor-id="beginner---intermediate">Beginner - Intermediate</h4>
<ul>
<li><a href="https://r4ds.had.co.nz/">R for Data Science</a></li>
</ul>
</section>
<section id="advanced-topics" class="level4">
<h4 class="anchored" data-anchor-id="advanced-topics">Advanced topics</h4>
<ul>
<li><p><a href="https://adv-r.hadley.nz/">Advanced R</a></p></li>
<li><p><a href="https://r-pkgs.org/">R Packages</a></p></li>
<li><p><a href="https://rmarkdown.rstudio.com/lesson-1.html">R Markdown lessons</a></p></li>
<li><p><a href="https://shiny.rstudio.com/tutorial/">Build Shiny applications</a></p></li>
</ul>
</section>
</section>
<section id="programming-in-python" class="level2">
<h2 class="anchored" data-anchor-id="programming-in-python">Programming in Python</h2>
<p>Python is a general-purpose programming language that is becoming increasingly popular with data scientists. Python has a growing number of libraries for data manipulation, exploration and visualisation as well as machine learning and is ideal for deploying applications and reproducible code. <a href="https://jupyter.org/">Jupyter</a> notebooks provide an interactive development environment for coding in Python and other languages.</p>
<section id="beginner---intermediate-1" class="level4">
<h4 class="anchored" data-anchor-id="beginner---intermediate-1">Beginner - Intermediate</h4>
<ul>
<li><p><a href="https://jakevdp.github.io/PythonDataScienceHandbook/">Python data science handbook</a></p>
<ul>
<li><a href="https://www.youtube.com/watch?v=_ZEWDGpM-vM&amp;list=PLYCpMb24GpOC704uO9svUrihl-HY1tTJJ">Accompanying videos</a></li>
</ul></li>
<li><p><a href="http://swcarpentry.github.io/python-novice-inflammation/">Programming with Python</a></p></li>
</ul>
</section>
<section id="advanced-topics-1" class="level4">
<h4 class="anchored" data-anchor-id="advanced-topics-1">Advanced topics</h4>
<ul>
<li><p>Building pipelines with SnakeMake</p>
<ul>
<li><p><a href="https://snakemake.readthedocs.io/en/stable/tutorial/tutorial.html">SnakeMake Tutorial</a></p></li>
<li><p><a href="https://slides.com/johanneskoester/snakemake-short#/">SnakeMake slides</a></p></li>
</ul></li>
</ul>
</section>
</section>
<section id="statistics-for-biology" class="level2">
<h2 class="anchored" data-anchor-id="statistics-for-biology">Statistics for biology</h2>
<p>While there are countless resources for learning statistics and a wealth of engaging videos on Youtube, it is important for bioinformaticians to understand statistics in the context of biology. Knowledge of statistical analysis will allow you to correctly describe and make inferences from your data as well as designing meaningful and useful experiments. Here are some links to get started:</p>
<ul>
<li><p><a href="https://docs.google.com/presentation/d/1qmtT3W6oylWs9xxeJ0WsNHyHbK6TiUJQsEQtFFVpMtU/edit?usp=sharing">Introduction to Stats</a> (slides)</p></li>
<li><p><a href="https://www.bioinformatics.babraham.ac.uk/training/R_Statistics/Introduction%20to%20Statistics%20with%20R%20manual.pdf">Introduction to statistics</a> with R</p></li>
<li><p><a href="https://www.nature.com/collections/qghhqm/pointsofsignificance">Points of Significance</a> (Nature collection)</p></li>
<li><p><a href="https://www.nature.com/collections/qghhqm">Statistics for biologists</a> (Nature collection)</p></li>
<li><p><a href="https://drive.google.com/open?id=1e9X5Tc4F8meqYWJOZn-Y2R7BcmjbQVg_">Statistical learning</a> (comprehensive stats applied in R)</p></li>
<li><p><a href="https://media.ed.ac.uk/channel/The+Roslin+Institute+-+Training/40716141">Roslin Institute Video Tutorials</a></p></li>
</ul>
</section>
<section id="analysing-hts-experiments" class="level2">
<h2 class="anchored" data-anchor-id="analysing-hts-experiments">Analysing HTS experiments</h2>
<p>The tutorials below provide introductions to analysing specific types of high throughput sequencing experiments. They are designed to help you understand your data, complete generic processing steps and explore and interpret the output. Please note the pre-requisite skills in brackets.</p>
<p>In most cases, the pipelines provided are simple and generic and may not be suitable for your own analysis. The bioinformatics core facility is available to collaborate on your project and to offer advice on experimental design and data analysis strategies. We also provide an expansive set of software tools and analysis workflows accessible through our bioinformatics servers:</p>
<section id="chip-seq" class="level4">
<h4 class="anchored" data-anchor-id="chip-seq">ChIP-seq</h4>
<ul>
<li><a href="https://danrobertson87.github.io/2020_chip-seq_workshop.github.io">WCB ChIP-seq analysis workshop</a> (Linux command line)</li>
</ul>
</section>
<section id="rna-seq" class="level4">
<h4 class="anchored" data-anchor-id="rna-seq">RNA-seq</h4>
<ul>
<li><p><a href="https://swebb1.github.io/RNA-seq_workshop">WCB RNA-seq analysis</a> (Linux command line)</p></li>
<li><p><a href="https://www.bioconductor.org/packages/devel/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html">Differential expression with DESeq2</a> (R)</p></li>
</ul>
</section>
<section id="hi-c" class="level4">
<h4 class="anchored" data-anchor-id="hi-c">Hi-C</h4>
<ul>
<li><a href="https://danrobertson87.github.io/2022_hic_workshop.github.io">WCB Hi-C data analysis</a> (command line)</li>
</ul>
</section>
<section id="clipcrac" class="level4">
<h4 class="anchored" data-anchor-id="clipcrac">CLIP/CRAC</h4>
<ul>
<li><a href="https://bitbucket.org/sgrann/pycrac/src/default/The%20pyCRAC%20Manual.pdf">PyCRAC manual</a></li>
</ul>
</section>
<section id="clash" class="level4">
<h4 class="anchored" data-anchor-id="clash">CLASH</h4>
<ul>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3969109/">Hyb methods paper</a></li>
</ul>
</section>
<section id="methyl-seq" class="level4">
<h4 class="anchored" data-anchor-id="methyl-seq">Methyl-seq</h4>
<ul>
<li><a href="https://www.bioinformatics.babraham.ac.uk/training/Methylation_Course/Basic%20BS-Seq%20processing%20Exercises.pdf">QC and alignment of BS-seq data using Bismark</a> (command line)</li>
</ul>
</section>
</section>
<section id="further-bioinformatics-training-resources" class="level2">
<h2 class="anchored" data-anchor-id="further-bioinformatics-training-resources">Further Bioinformatics Training Resources</h2>
<p><a href="https://carpentries.org/">Carpentries</a></p>
<ul>
<li>Community driven project delivering training in data science and coding skills. <a href="https://software-carpentry.org/">Software Carpentry</a> and <a href="http://www.datacarpentry.org/">Data Carpentry</a> websites include many free online courses. <a href="https://edcarp.github.io/">Edinburgh Carpentries</a> host regular events at the university and include a Data Carpentry for Genomics course.</li>
</ul>
<p><a href="http://rafalab.github.io/pages/harvardx.html">HarvardX</a></p>
<ul>
<li>A comprehensive set of courses covering R, Python, and biostatistics among other things. Includes case studies in analysing RNA-seq, ChIP-seq and Methyl-seq data.</li>
</ul>
<p><a href="https://github.com/hbctraining">Harvard Chan Bioinformatics Core</a></p>
<ul>
<li>Training courses in R and command line including many different HTS techniques.</li>
</ul>
<p><a href="https://www.bioinformatics.babraham.ac.uk/training.html">Babraham Bioinformatics</a></p>
<ul>
<li><p>Training courses offered by the Babraham Bioinformatics group in R, Python, statistics and sequencing applications.</p></li>
<li><p>Bitesize bioinformatics <a href="https://tinyurl.com/bitebioinf">modules</a> and <a href="https://www.youtube.com/playlist?list=PLbiByRpDb_hP7b-I1GR4eEWCD2OqdZEg1">videos</a></p></li>
</ul>
<p><a href="https://rockefelleruniversity.github.io/training.html">Rockefeller Bioinformatics</a></p>
<ul>
<li>A collection of bioinformatics training courses</li>
</ul>
<p><a href="https://biodatascience.github.io/">Biodatascience</a></p>
<ul>
<li>A collection of R courses in data science, statistics and computational biology.</li>
</ul>
<p><a href="http://www.datacamp.com/">Datacamp</a></p>
<ul>
<li>Interactive learning with videos, lessons, tests and a built-in programming interface. Datacamp is a subscription service with some free modules for learning R, Python and data science skills.</li>
</ul>
<p><a href="https://github.com/bioinformatics-core-shared-training">UoC bioinformatics core</a></p>
<ul>
<li>A collection of bioinformatics training courses from the University of Cambridge bioinformatics core facility.</li>
</ul>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
          // target, if specified
          link.setAttribute("target", "_blank");
          if (link.getAttribute("rel") === null) {
            link.setAttribute("rel", "noopener");
          }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>